### PR TITLE
Example snippet to query by protocolPath and structure

### PR DESCRIPTION
**Pull request by blackgirlbytes**
## What type of PR is this? (check all applicable)

- [ ] ♻️ Refactor
- [ ] ✨ New Feature
- [ ] 🐛 Bug Fix
- [x] 📝 Documentation Update
- [ ] 👷 Example Application
- [ ] 🧑‍💻 Code Snippet
- [ ] 🎨 Design
- [ ] 📖 Content
- [ ] 🧪 Tests
- [ ] 🔖 Release
- [ ] 🚩 Other

## Description

This PR adds an example of how to query records by protocolPath and type. I saw someone ask about this in Discord, and I feel like it would help to have a more explicit example.

I got Nick's review, so I guess I only need a review for clarity and structure. 

## Related Tickets & Documents

Resolves #

## Mobile & Desktop Screenshots/Recordings

Screenshot 2023-09-29 at 9 59 21 AM

## Added code snippets?
- [x] 👍 yes
- [ ] 🙅 no, because they aren't needed

## Added tests?

- [x] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help (will add)

## Added to documentation?

- [ ] 📜 readme
- [ ] 📜 contributing.md
- [x] 📓 general documentation
- [ ] 🙅 no documentation needed

## [optional] What gif best describes this PR or how it makes you feel?
giphy

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1205620813630998

### CODE REVIEW

File: site/docs/web5/build/decentralized-web-nodes/query-from-dwn.mdx [hunk 1]
```diff
@@ -25,6 +25,30 @@ The following snippet queries a DWN for records that were written by a specific
 To guarantee record data, call [records.read()](/api/web5-js/dwn/records#readrequest).
 :::
 
+## Querying Records by Path and Type
+The following snippets demonstrate how to query records by path and type to match a defined protocol. The protocol definition below has an audio type and a video type:
+
+```javascript
+  const musicProtocolDefinition = {
```
Comment (angiejones) [inline site/docs/web5/build/decentralized-web-nodes/query-from-dwn.mdx:32]: this part should also go in a snippet. For example, they added `published` as a required field on protocols and we didn't know - and our code snippets didn't reflect that. this is the problem we're trying to solve for with snippets and tests.

File: site/docs/web5/build/decentralized-web-nodes/query-from-dwn.mdx [hunk 2]
```diff
@@ -25,6 +25,30 @@ The following snippet queries a DWN for records that were written by a specific
 To guarantee record data, call [records.read()](/api/web5-js/dwn/records#readrequest).
 :::
 
+## Querying Records by Path and Type
+The following snippets demonstrate how to query records by path and type to match a defined protocol. The protocol definition below has an audio type and a video type:
```
Comment (angiejones) [inline site/docs/web5/build/decentralized-web-nodes/query-from-dwn.mdx:29]:
```suggestion
The following snippets demonstrate how to query records by protocol path and type to match a defined protocol. The protocol definition below has an audio type and a video type:
```

Comment (angiejones) [reply site/docs/web5/build/decentralized-web-nodes/query-from-dwn.mdx:29]: Not sure if you're actually querying by `type`. I think it's actually on structure. 

For example, in this example, my protocolPath is `protocolPath: 'message/reply',` (not just `reply` - that doesn't work), where the protocol is:

```js
const protocolDefinition = {
    "protocol": "https://social-media.xyz",
    "published": true,
    "types": {
      "message": {
        "schema": "https://social-media.xyz/schemas/messageSchema",
        "dataFormats": ["text/plain"]
      },
      "reply": {
        "schema": "https://social-media.xyz/schemas/replySchema",
        "dataFormats": ["text/plain"]
      },
      "image": {
        "dataFormats": ["image/jpeg"]
      },
      "caption": {
        "schema": "https://social-media.xyz/schemas/captionSchema",
        "dataFormats": ["text/plain"]
      }
    },
    "structure": {
      "message": {
        "$actions": [...],
        "reply": {
          "$actions": [...]
        }
      },
      "image": {
        "$actions": [...],
        "caption": {
          "$actions": [...]
        },
        "reply": {
          "$actions": [...]
        }
      }
    }
};
```

**Review by angiejones — CHANGES_REQUESTED** [review]

**Comment by blackgirlbytes** [comment]
I revised it based on your feedback @angiejones

File: site/code-snippets/web5/build/decentralized-web-nodes/query-from-dwn.js [hunk 3]
```diff
@@ -64,7 +64,40 @@ export async function queryRecordWithParentId(web5) {
...
+        schema: "https://schema.org/Playlist",
+        dataFormats: ["audio/mp3"],
+      },
+      video: {
+        schema: "https://schema.org/Playlist",
+        dataFormats: ["video/mp4"],
+      },
+    },
+    structure: {
+      // highlight-start
```
Comment (angiejones) [inline site/code-snippets/web5/build/decentralized-web-nodes/query-from-dwn.js:82]: no need to change this time, but just wanted to let you know there's a `// highlight-next-line` for when you just need to highlight a single line 😊

File: site/code-snippets/web5/build/decentralized-web-nodes/query-from-dwn.js [hunk 4]
```diff
@@ -64,7 +64,40 @@ export async function queryRecordWithParentId(web5) {
...
 }
 
-export async function queryFromDwnByPathAndType(web5) {
+export async function queryFromDwnByPathAndStructure(web5) {
+
+  const musicProtocolDefinition = {
+    protocol: "https://music.org/protocol",
+    types: {
+      audio: {
+        schema: "https://schema.org/Playlist",
```
Comment (angiejones) [inline site/code-snippets/web5/build/decentralized-web-nodes/query-from-dwn.js:73]: if you'd like to use a resolvable schema, can use https://schema.org/AudioObject

File: site/code-snippets/web5/build/decentralized-web-nodes/query-from-dwn.js [hunk 5]
```diff
@@ -64,7 +64,40 @@ export async function queryRecordWithParentId(web5) {
...
+        dataFormats: ["audio/mp3"],
+      },
+      video: {
+        schema: "https://schema.org/Playlist",
```
Comment (angiejones) [inline site/code-snippets/web5/build/decentralized-web-nodes/query-from-dwn.js:77]: https://schema.org/VideoObject

File: site/code-snippets/web5/build/decentralized-web-nodes/query-from-dwn.js [hunk 6]
```diff
@@ -64,7 +64,40 @@ export async function queryRecordWithParentId(web5) {
   return response;
 }
 
-export async function queryFromDwnByPathAndType(web5) {
+export async function queryFromDwnByPathAndStructure(web5) {
+
+  const musicProtocolDefinition = {
+    protocol: "https://music.org/protocol",
```
Comment (angiejones) [inline site/code-snippets/web5/build/decentralized-web-nodes/query-from-dwn.js:70]:
```suggestion
    protocol: "https://music.org/protocol",
    published:true,
```
The protocol won't install without the `published` property being defined

**Review by angiejones — CHANGES_REQUESTED** [review]

File: site/code-snippets/web5/build/decentralized-web-nodes/query-from-dwn.js [hunk 7]
```diff
@@ -63,3 +63,51 @@ export async function queryRecordWithParentId(web5) {
...
+    types: {
+      audio: {
+        schema: "https://schema.org/Playlist",
+        dataFormats: ["audio/mp3"],
+      },
+      video: {
+        schema: "https://schema.org/Playlist",
```
Comment (angiejones) [inline site/code-snippets/web5/build/decentralized-web-nodes/query-from-dwn.js:77]:
```suggestion
        schema: "https://schema.org/VideoObject",
```

File: site/code-snippets/web5/build/decentralized-web-nodes/query-from-dwn.js [hunk 8]
```diff
@@ -63,3 +63,51 @@ export async function queryRecordWithParentId(web5) {
...
   return response;
 }
+
+export async function queryFromDwnByPathAndStructure(web5) {
+
+  const musicProtocolDefinition = {
+    protocol: "https://music.org/protocol",
+    types: {
+      audio: {
+        schema: "https://schema.org/Playlist",
```
Comment (angiejones) [inline site/code-snippets/web5/build/decentralized-web-nodes/query-from-dwn.js:73]:
```suggestion
        schema: "https://schema.org/AudioObject",
```

if you'd like to use a resolvable schema, can use https://schema.org/AudioObject

File: site/code-snippets/web5/build/decentralized-web-nodes/query-from-dwn.js [hunk 9]
```diff
@@ -63,3 +63,51 @@ export async function queryRecordWithParentId(web5) {
...
+      }
+    }
+  };
+
+  const { records } = await web5.dwn.records.query({
+    message: {
+      filter: {
+        protocol: "https://example.com/musicProtocol",
+        // highlight-start
+        protocolPath: 'audio',
```
Comment (angiejones) [inline site/code-snippets/web5/build/decentralized-web-nodes/query-from-dwn.js:106]: do you think it's worth showing an example where you're not querying a root level object? For example, if your structure was defined as a Playlist that can include types Audio and Video, then the protocolPath would be `playlist/audio`. Not sure if that's obvious to folks.

File: site/code-snippets/web5/build/decentralized-web-nodes/query-from-dwn.js [hunk 10]
```diff
@@ -80,32 +84,37 @@ export async function queryFromDwnByPathAndStructure(web5) {
       },
     },
     structure: {
-      // highlight-start
-      audio: {
-        // highlight-end
+      playlist: {
```
Comment (angiejones) [inline site/code-snippets/web5/build/decentralized-web-nodes/query-from-dwn.js:87]: who writes a playlist? did you test to see if the author can write a new one to their DWN based on this protocol?

File: site/docs/web5/build/decentralized-web-nodes/query-from-dwn.mdx [hunk 11]
```diff
@@ -25,6 +25,15 @@ The following snippet queries a DWN for records that were written by a specific
 To guarantee record data, call [records.read()](/api/web5-js/dwn/records#readrequest).
 :::
 
+## Querying Records by Path and Structure
```
Comment (angiejones) [inline site/docs/web5/build/decentralized-web-nodes/query-from-dwn.mdx:28]:
```suggestion
## Querying Records by Protocol Path
```

File: site/docs/web5/build/decentralized-web-nodes/query-from-dwn.mdx [hunk 12]
```diff
@@ -25,6 +25,15 @@ The following snippet queries a DWN for records that were written by a specific
...
+
+The following snippet demonstrates how to query records by protocol path and structure to match a defined protocol:
```
Comment (angiejones) [inline site/docs/web5/build/decentralized-web-nodes/query-from-dwn.mdx:30]:
```suggestion
The following snippet demonstrates how to query records that match a specific protocol path:
```

**Review by angiejones — CHANGES_REQUESTED** [review]

**Review by angiejones — APPROVED** [review]

File: site/code-snippets/web5/build/decentralized-web-nodes/query-from-dwn.js (hunk 10 above)
Comment (blackgirlbytes) [reply site/code-snippets/web5/build/decentralized-web-nodes/query-from-dwn.js:87]: this is a similar structure i have for the dinger app. And it was working for me. Is the wrong way to set permissions? I thought this was saying:
Anyone can create a playlist.
If an author creates a playlist, they can read their playlist.
They can send a playlist to someone and that recipient can read their playlist.

Am I misunderstanding?

Comment (angiejones) [reply site/code-snippets/web5/build/decentralized-web-nodes/query-from-dwn.js:87]: with the `anyone` can write permission, then yes. I think that was missing at first so no one would have been able to write a playlist object

```js
        $actions: [
          { who: "anyone", can: "write" },
          { who: "author", of: "playlist", can: "read" },
          { who: "recipient", of: "playlist", can: "read" },
        ],
```

it's correct now 😊

File: site/code-snippets/web5/build/decentralized-web-nodes/query-from-dwn.js [hunk 13]
```diff
@@ -63,3 +63,61 @@ export async function queryRecordWithParentId(web5) {
...
+      },
+    }
+  };
+
+  const { records } = await web5.dwn.records.query({
+    message: {
+      filter: {
+        protocol: "https://playlist.org/protocol",
+        // highlight-next-line
+        protocolPath: 'playlist/videos',
```
Comment (angiejones) [inline site/code-snippets/web5/build/decentralized-web-nodes/query-from-dwn.js:117]:
```suggestion
        protocolPath: 'playlist/video',
```

**Review by angiejones — CHANGES_REQUESTED** [review]

File: site/docs/web5/build/decentralized-web-nodes/query-from-dwn.mdx [hunk 14]
```diff
@@ -25,6 +25,15 @@ The following snippet queries a DWN for records that were written by a specific
...
 :::
 
+## Querying Records by Protocol Path
+
+The following snippet demonstrates how to query records that match a specific protocol path:
+
+<CodeSnippet functionName="queryFromDwnByPathAndStructure" />
+
+:::note
+Before querying records by path and structure, ensure that the protocol is [defined](https://developer.tbd.website/docs/web5/learn/protocols/#defining-a-protocol) and [installed](https://developer.tbd.website/docs/web5/learn/protocols/#installing-a-protocol).
```
Comment (angiejones) [inline site/docs/web5/build/decentralized-web-nodes/query-from-dwn.mdx:35]:
```suggestion
Before querying records by protocol path, ensure that the protocol is [installed](https://developer.tbd.website/docs/web5/learn/protocols/#installing-a-protocol). 
```

**Review by angiejones — CHANGES_REQUESTED** [review]

File: site/code-snippets/web5/build/decentralized-web-nodes/query-from-dwn.js (hunk 13 above)
Comment (angiejones) [reply site/code-snippets/web5/build/decentralized-web-nodes/query-from-dwn.js:117]: Always run any code you put out to make sure it actually works. I just ran all the code to make sure it works, and it's good to go (with this change):
- installed this protocol
- created a playlist record
- created a video record that's a child of playlist
- ran the query

Comment (blackgirlbytes) [reply site/code-snippets/web5/build/decentralized-web-nodes/query-from-dwn.js]: true after I made a bunch of changes, I did neglect to test it! Thanks!

**Review by angiejones — APPROVED** [review]

**Pull request merged**